File: frontend/src/types/weather.ts
export interface WeatherData {
	location: Coordinates;
	timestamp: Date;
	temperature: number;
	humidity: number;
	rainfall: number;
	windSpeed: number;
	windDirection: number;
	pressure: number;
	uvIndex: number;
	elevation: number; // Floor level or ground elevation
}

export interface Coordinates {
	latitude: number;
	longitude: number;
	elevation?: number; // Meters above sea level or floor level
}

export interface MicroclimateGrid {
	bounds: {
		minLat: number;
		maxLat: number;
		minLng: number;
		maxLng: number;
	};
	resolution: number; // Grid size in meters (e.g., 100m x 100m)
	data: GridCell[];
}

export interface GridCell {
	coordinates: Coordinates;
	weather: WeatherData;
	confidence: number; // 0-1, based on sensor density
	source: 'official' | 'crowdsourced' | 'interpolated' | 'ml-predicted';
}

export interface VerticalWeatherProfile {
	location: Coordinates;
	timestamp: Date;
	layers: WeatherLayer[];
}

export interface WeatherLayer {
	elevationRange: [number, number]; // [minFloor, maxFloor] or [minElevation, maxElevation]
	temperature: number;
	humidity: number;
	visibility: number; // For fog prediction
	rainfall: number;
	windSpeed: number;
}

export interface Alert {
	id: string;
	type: 'typhoon' | 'rainstorm' | 'heat' | 'cold' | 'custom';
	severity: 'info' | 'warning' | 'danger';
	title: string;
	message: string;
	affectedArea: {
		type: 'point' | 'radius' | 'polygon';
		coordinates: Coordinates | Coordinates[];
		radius?: number; // For radius type
	};
	validFrom: Date;
	validUntil: Date;
}

export interface SensorReading {
	sensorId: string;
	userId?: string;
	location: Coordinates;
	timestamp: Date;
	readings: {
		temperature?: number;
		humidity?: number;
		pressure?: number;
		rainfall?: number;
	};
	deviceType: 'bluetooth-thermometer' | 'smartphone-barometer' | 'weather-station' | 'other';
	accuracy: number; // Confidence score 0-1
}

export interface LaundryIndex {
	location: Coordinates;
	timestamp: Date;
	dryTimeMinutes: number;
	mouldRiskScore: number; // 0-100
	recommendation: 'excellent' | 'good' | 'fair' | 'poor' | 'avoid';
	factors: {
		temperature: number;
		humidity: number;
		sunlight: number; // 0-1
		windSpeed: number;
		buildingShadow: boolean;
	};
}

export interface BuildingData {
	buildingId: string;
	address: string;
	coordinates: Coordinates;
	height: number; // Meters
	floors: number;
	facing: 'north' | 'south' | 'east' | 'west' | 'mixed';
	surroundingBuildings: BuildingShadow[];
}

export interface BuildingShadow {
	buildingId: string;
	direction: number; // Degrees from north
	height: number;
	distance: number; // Meters
}

export interface ForecastModel {
	modelId: string;
	version: string;
	type: 'official' | 'ml-corrected' | 'crowd-sourced';
	accuracy: number;
	validFor: {
		region: 'hong-kong-island' | 'kowloon' | 'new-territories' | 'all';
		timeHorizon: number; // Hours ahead
	};
}

export interface OfflineData {
	location: Coordinates;
	radius: number; // Meters
	downloadedAt: Date;
	expiresAt: Date;
	weatherData: MicroclimateGrid;
	forecastModel: ArrayBuffer; // TensorFlow Lite model
	mapTiles: string[]; // URLs or base64 encoded tiles
	alerts: Alert[];
}

export type WeatherLayer3D = 'temperature' | 'humidity' | 'rainfall' | 'wind' | 'pressure' | 'laundry-index' | 'mould-risk';

export interface MapViewState {
	longitude: number;
	latitude: number;
	zoom: number;
	pitch: number;
	bearing: number;
	minZoom: number;
	maxZoom: number;
}

File: backend-api/app/schemas/weather.py
from pydantic import BaseModel, Field
from typing import Optional, List, Literal
from datetime import datetime


class Coordinates(BaseModel):
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    elevation: Optional[float] = Field(None, ge=0)


class WeatherResponse(BaseModel):
    location: Coordinates
    timestamp: datetime
    temperature: float
    humidity: float
    rainfall: float
    wind_speed: float
    wind_direction: Optional[float]
    pressure: Optional[float]
    uv_index: Optional[float]
    elevation: float


class GridBounds(BaseModel):
    min_lat: float = Field(..., alias="minLat")
    max_lat: float = Field(..., alias="maxLat")
    min_lng: float = Field(..., alias="minLng")
    max_lng: float = Field(..., alias="maxLng")

    class Config:
        populate_by_name = True


class WeatherGridRequest(BaseModel):
    bounds: GridBounds
    resolution: int = Field(100, ge=50, le=500)  # Grid size in meters


class GridCell(BaseModel):
    coordinates: Coordinates
    weather: WeatherResponse
    confidence: float = Field(..., ge=0, le=1)
    source: Literal["official", "crowdsourced", "interpolated", "ml-predicted"]


class WeatherGridResponse(BaseModel):
    bounds: GridBounds
    resolution: int
    data: List[GridCell]


class WeatherLayer(BaseModel):
    elevation_range: tuple[float, float]
    temperature: float
    humidity: float
    visibility: float
    rainfall: float
    wind_speed: float


class VerticalProfileResponse(BaseModel):
    location: Coordinates
    timestamp: datetime
    layers: List[WeatherLayer]


class LaundryIndexResponse(BaseModel):
    location: Coordinates
    timestamp: datetime
    dry_time_minutes: int
    mould_risk_score: float = Field(..., ge=0, le=100)
    recommendation: Literal["excellent", "good", "fair", "poor", "avoid"]
    factors: dict


class SensorReading(BaseModel):
    sensor_id: str
    location: Coordinates
    timestamp: datetime
    readings: dict
    device_type: str
    accuracy: float = Field(..., ge=0, le=1)


class AlertCreate(BaseModel):
    alert_type: Literal["typhoon", "rainstorm", "heat", "cold", "custom"]
    severity: Literal["info", "warning", "danger"]
    title: str
    message: str
    affected_area: dict
    valid_from: datetime
    valid_until: datetime


class AlertResponse(BaseModel):
    id: str
    alert_type: str
    severity: str
    title: str
    message: str
    affected_area: dict
    valid_from: datetime
    valid_until: datetime
    created_at: datetime
